File: dev/requirements/logic/cases/countdown-tone.2.8.case.swift
import Foundation
import SetlistCore

/// 2.8 — Neutral above three minutes remaining, orange at three minutes or
/// less, red at zero, counting up as a negative after that.
let countdownTone_2_8 = RequirementCase(id: "2.8") {
    try expectEqual(Countdown.remaining(plannedLength: 1200, elapsed: 1122), 78, "remaining time")
    try expectEqual(Countdown.remaining(plannedLength: 1200, elapsed: 1260), -60, "remaining time past the slot")

    try expectEqual(Countdown.tone(remaining: 600), .neutral, "ten minutes left")
    try expectEqual(Countdown.tone(remaining: 181), .neutral, "just above the threshold")
    try expectEqual(Countdown.tone(remaining: 180), .warning, "exactly at the threshold")
    try expectEqual(Countdown.tone(remaining: 78), .warning, "inside the last three minutes")
    try expectEqual(Countdown.tone(remaining: 0), .over, "on the buzzer")
    try expectEqual(Countdown.tone(remaining: -60), .over, "a minute over")
}

File: dev/requirements/logic/cases/countdown-tone.2.8.case.dart
import 'package:setlist_core/setlist_core.dart';

import '../../shared/expectations.dart';
import '../../shared/requirement_case.dart';

/// 2.8 — The countdown is neutral above three minutes remaining, orange at
/// three minutes or less, red once remaining time reaches zero.
final countdownTone_2_8 = RequirementCase('2.8', (tester) {
  expectThat(
    Countdown.remaining(
      plannedLength: const Duration(minutes: 20),
      elapsed: const Duration(minutes: 18, seconds: 42),
    ),
    const Duration(seconds: 78),
    'remaining time',
  );
  expectThat(
    Countdown.remaining(
      plannedLength: const Duration(minutes: 20),
      elapsed: const Duration(minutes: 21),
    ),
    const Duration(minutes: -1),
    'remaining time past the slot counts up as a negative',
  );

  expectThat(Countdown.tone(const Duration(minutes: 10)), CountdownTone.neutral, 'ten minutes left');
  expectThat(
    Countdown.tone(const Duration(minutes: 3, seconds: 1)),
    CountdownTone.neutral,
    'just above the threshold',
  );
  expectThat(
    Countdown.tone(const Duration(minutes: 3)),
    CountdownTone.warning,
    'exactly at the threshold',
  );
  expectThat(
    Countdown.tone(const Duration(seconds: 78)),
    CountdownTone.warning,
    'inside the last three minutes',
  );
  expectThat(Countdown.tone(Duration.zero), CountdownTone.over, 'on the buzzer');
  expectThat(Countdown.tone(const Duration(minutes: -1)), CountdownTone.over, 'a minute over');
});
